// @oracle/bots-node-sdk/typings/lib.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Lib = void 0;
const lib = require("./lib/"); // index
/**
 * SDK 'Lib' exposing core classes, decorators, etc.
 */
var Lib;
(function (Lib) {
    // Custom component @Component decorator
    // export const Component = lib.Component;
    // export type Component = lib.Component;
    // Legacy Shell (for use by swagger)
    Lib.ComponentShell = lib.ComponentShell;
    // Custom Component abstraction class
    class ComponentAbstract extends lib.ComponentAbstract {
    }
    Lib.ComponentAbstract = ComponentAbstract;
    // Custom Component registry class
    class ComponentRegistry extends lib.ComponentRegistry {
    }
    Lib.ComponentRegistry = ComponentRegistry;
    // Component Conversation invocation
    class CustomComponentContext extends lib.CustomComponentContext {
    }
    Lib.CustomComponentContext = CustomComponentContext;
    // Entity Resolution invocation
    class EntityResolutionContext extends lib.EntityResolutionContext {
    }
    Lib.EntityResolutionContext = EntityResolutionContext;
    // Conversation Message Model
    class MessageModel extends lib.MessageModel {
    }
    Lib.MessageModel = MessageModel;
})(Lib = exports.Lib || (exports.Lib = {}));
