// @oracle/bots-node-sdk/typings/lib2/messagev2/common/paginationInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaginationInfo = void 0;
/**
 * Represents pagination information.
 */
class PaginationInfo {
    /**
     * Creates an instance of the PaginationInfo class.
     * @param {number} totalCount The total count.
     * @param {number} rangeSize The range size.
     * @param {number} rangeStart The range start.
     * @param {string} status The status.
     */
    constructor(totalCount, rangeSize, rangeStart) {
        this.totalCount = totalCount;
        this.rangeSize = rangeSize;
        this.rangeStart = rangeStart;
    }
    /**
     * Gets the total count.
     * @returns {number} The total count.
     */
    getTotalCount() {
        return this.totalCount;
    }
    /**
     * Sets the total count.
     * @param {number} totalCount The total count to set.
     * @returns The updated instance of the PaginationInfo.
     */
    setTotalCount(totalCount) {
        this.totalCount = totalCount;
        return this;
    }
    /**
     * Gets the range size.
     * @returns {number} The range size.
     */
    getRangeSize() {
        return this.rangeSize;
    }
    /**
     * Sets the range size.
     * @param {number} rangeSize The range size to set.
     * @returns The updated instance of the PaginationInfo.
     */
    setRangeSize(rangeSize) {
        this.rangeSize = rangeSize;
        return this;
    }
    /**
     * Gets the range start.
     * @returns {number} The range start.
     */
    getRangeStart() {
        return this.rangeStart;
    }
    /**
     * Sets the range start.
     * @param {number} rangeStart The range start to set.
     * @returns The updated instance of the PaginationInfo.
     */
    setRangeStart(rangeStart) {
        this.rangeStart = rangeStart;
        return this;
    }
    /**
     * Gets the status.
     * @returns {string} The status.
     */
    getStatus() {
        return this.status;
    }
    /**
     * Sets the status.
     * @param {string} status The status to set.
     * @returns The updated instance of the PaginationInfo.
     */
    setStatus(status) {
        this.status = status;
        return this;
    }
}
exports.PaginationInfo = PaginationInfo;
