// @oracle/bots-node-sdk/typings/lib2/messagev2/common/voice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Voice = void 0;
/**
 * Represents the voice settings of a message element.
 */
class Voice {
    /**
     * Constructs a Voice object with the specified text.
     * @param {string} text The text of the voice (required).
     */
    constructor(text) {
        this.text = text;
    }
    /**
     * Gets the text of the voice.
     * @returns {string} The text of the voice.
     */
    getText() {
        return this.text;
    }
    /**
     * Sets the text of the voice.
     * @param {string} text The text of the voice.
     * @returns {Voice} This Voice instance.
     */
    setText(text) {
        this.text = text;
        return this;
    }
    /**
     * Gets the sound URL of the voice.
     * @returns {string} The sound URL of the voice.
     */
    getSoundUrl() {
        return this.soundUrl;
    }
    /**
     * Sets the sound URL of the voice.
     * @param {string} soundUrl The sound URL of the voice.
     * @returns {Voice} This Voice instance.
     */
    setSoundUrl(soundUrl) {
        this.soundUrl = soundUrl;
        return this;
    }
    /**
     * Gets the long text of the voice.
     * @returns {string} The long text of the voice.
     */
    getLongText() {
        return this.longText;
    }
    /**
     * Sets the long text of the voice.
     * @param {string} longText The long text of the voice.
     * @returns {Voice} This Voice instance.
     */
    setLongText(longText) {
        this.longText = longText;
        return this;
    }
}
exports.Voice = Voice;
