// @oracle/bots-node-sdk/typings/lib2/messagev2/field/readOnlyField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReadOnlyField = void 0;
const internal_1 = require("../internal");
/**
 * Represents a read-only field.
 * @extends Field
 */
class ReadOnlyField extends internal_1.Field {
    /**
     * Gets the value of the read-only field.
     * @returns {any} The value of the field.
     */
    getValue() {
        return this.value;
    }
    /**
     * Sets the value of the read-only field.
     * @param {any} value The value to set.
     * @returns The current instance of the ReadOnlyField class.
     */
    setValue(value) {
        this.value = value;
        return this;
    }
    /**
     * Gets the width of the read-only field.
     * @returns {number} The width value.
     */
    getWidth() {
        return this.width;
    }
    /**
     * Sets the width of the read-only field.
     * @param {number} width The width value to set.
     * @returns The current instance of the ReadOnlyField class.
     */
    setWidth(width) {
        this.width = width;
        return this;
    }
    /**
     * Gets the alignment of the read-only field.
     * @returns {FieldAlignment} The alignment value.
     */
    getAlignment() {
        return this.alignment;
    }
    /**
     * Sets the alignment of the read-only field.
     * @param {FieldAlignment} alignment The alignment value to set.
     * @returns The current instance of the ReadOnlyField class.
     */
    setAlignment(alignment) {
        this.alignment = alignment;
        return this;
    }
}
exports.ReadOnlyField = ReadOnlyField;
